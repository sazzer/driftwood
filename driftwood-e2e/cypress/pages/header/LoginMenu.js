/// <reference types="cypress" />

export default class LoginMenu {
    constructor(element) {
        this._element = element;
        element.should('not.have.class', 'loading');
    }

    assertIsOpen() {
        this._element.should('have.class', 'visible');
    }

    assertIsClosed() {
        this._element.should('not.have.class', 'visible');
    }

    assertProviderPresent(provider) {
        this._element.get(`div.menu div[data-provider="${provider}"]`);
    }

    open() {
        this._element.click();
    }
}
