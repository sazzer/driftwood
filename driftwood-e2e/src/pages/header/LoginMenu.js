const {By} = require('selenium-webdriver');

/**
 * Page Object representing the login menu
 */
class LoginMenu {
    /**
     * Construct the login menu
     * @param element the web element representing the login menu
     */
    constructor(element) {
        this._element = element;
    }

    /**
     * Check if the menu is open
     * @return True if the menu is open. False if not
     */
    async isOpen() {
        const expanded = await this._element.getAttribute('aria-expanded');
        return expanded === 'true';
    }

    /**
     * Open the menu
     */
    async open() {
        const isOpen = await this.isOpen();
        if (!isOpen) {
            await this._element.click();
        }
    }

    /**
     * Get the providers with which we can log in
     * @return The list of providers
     */
    async getProviders() {
        await this.open();

        const providerOptions = await this._element.findElements(By.css('[data-provider]'));
        const providers = providerOptions.map(async (provider) => await provider.getAttribute('data-provider'));

        return await Promise.all(providers);
    }

    /**
     * Log in with the given provider
     * @param providerName the name of the provider
     */
    async login(providerName) {
        await this.open();

        const provider = await this._element.findElement(By.css(`[data-provider="${providerName}"]`));
        await provider.click();
    }
}

module.exports = LoginMenu;
