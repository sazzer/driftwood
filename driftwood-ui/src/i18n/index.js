// @flow

import i18n from 'i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import defaultTranslations from './messages.json';

i18n
    .use(LanguageDetector)
    .init({
        resources: {
            dev: {
                driftwood: defaultTranslations
            }
        },

        // have a common namespace used around the full app
        ns: ['driftwood'],
        defaultNS: 'driftwood',

        debug: false,

        interpolation: {
            escapeValue: false
        },

        parseMissingKeyHandler: (key) => `!!${key}!!`
    });

export default i18n;
