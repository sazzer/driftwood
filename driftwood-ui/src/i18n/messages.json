{
  "page": {
    "title": "Driftwood"
  },
  "authentication": {
    "menu": {
      "title": "Log In",
      "loadingError": "Failed to load Log In options",
      "facebook" : "Facebook",
      "google": "Google",
      "twitter": "Twitter"
    }
  }
}
