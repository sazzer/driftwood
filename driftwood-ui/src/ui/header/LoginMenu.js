// @flow

import React from 'react';
import {NamespacesConsumer} from 'react-i18next';
import {Dropdown} from 'semantic-ui-react';
import LoginMenuItem from './LoginMenuItem';

/** The possible statuses of the Login Menu */
export const LoginMenuStatus = {
    loading: Symbol('loading'),
    loaded: Symbol('loaded'),
    failed: Symbol('failed'),
};

/**
 * The props for the Login Menu component
 */
export type LoginMenuProps = {
    providers: Array<string>,
    status?: Symbol,
}

/**
 * The component representing the login menu
 * @constructor
 */
export function LoginMenu({providers, status}: LoginMenuProps) {
    let menuItems;

    if (status === LoginMenuStatus.failed) {
        menuItems = (
            <NamespacesConsumer>
                {
                    (t) => <Dropdown.Item icon='ban' text={t(`authentication.menu.loadingError`)} />
                }
            </NamespacesConsumer>
        );
    } else {
        menuItems = providers.map(provider => <LoginMenuItem key={provider} provider={provider} />);
    }

    return (
        <NamespacesConsumer>
            {
                (t) => (
                    <Dropdown item
                              simple
                              text={t('authentication.menu.title')}
                              openOnFocus={false}
                              data-test='loginMenu'
                              loading={status === LoginMenuStatus.loading}
                              error={status === LoginMenuStatus.failed} >
                        <Dropdown.Menu>
                            { menuItems }
                        </Dropdown.Menu>
                    </Dropdown>
                )
            }
        </NamespacesConsumer>
    );
}

export default LoginMenu;
