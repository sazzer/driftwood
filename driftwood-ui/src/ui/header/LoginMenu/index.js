// @flow

import LoginMenu from './ConnectedLoginMenu';
export default LoginMenu;
