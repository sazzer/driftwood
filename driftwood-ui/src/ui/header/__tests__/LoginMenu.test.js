// @flow

import React from 'react';
import {render} from 'enzyme';
import LoginMenu from '../LoginMenu';
import i18n from "../../../i18n";
import {I18nextProvider} from "react-i18next";

/** Set up the component to test */
function setup(providers: Array<string>) {
    return {
        element: render(<I18nextProvider i18n={ i18n }><LoginMenu providers={providers} /></I18nextProvider>)
    };
}

it('Renders correctly with no providers', () => {
    const { element } = setup([]);

    expect(element).toMatchSnapshot();
});

it('Renders correctly with some providers', () => {
    const { element } = setup(['google', 'twitter']);

    expect(element).toMatchSnapshot();
});
