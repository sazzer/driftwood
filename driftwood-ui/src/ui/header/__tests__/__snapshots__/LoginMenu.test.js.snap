// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with no providers 1`] = `
<div
  aria-busy="false"
  aria-expanded="false"
  class="ui item simple dropdown"
  role="listbox"
  tabindex="0"
>
  <div
    aria-live="polite"
    class="text"
    role="alert"
  >
    Log In
  </div>
  <i
    aria-hidden="true"
    class="dropdown icon"
  />
  <div
    class="menu transition"
  />
</div>
`;

exports[`Renders correctly with some providers 1`] = `
<div
  aria-busy="false"
  aria-expanded="false"
  class="ui item simple dropdown"
  role="listbox"
  tabindex="0"
>
  <div
    aria-live="polite"
    class="text"
    role="alert"
  >
    Log In
  </div>
  <i
    aria-hidden="true"
    class="dropdown icon"
  />
  <div
    class="menu transition"
  >
    <div
      class="item"
      role="option"
    >
      <i
        aria-hidden="true"
        class="google icon"
      />
      <span
        class="text"
      >
        Google
      </span>
    </div>
    <div
      class="item"
      role="option"
    >
      <i
        aria-hidden="true"
        class="twitter icon"
      />
      <span
        class="text"
      >
        Twitter
      </span>
    </div>
  </div>
</div>
`;
